node/network/src/proxy/throttling.rs

```diff
@@ -0,0 +1,61 @@
+use std::time::{Duration, Instant};
+
+use futures::future::Future;
+use futures::sink::Sink;
+use futures::Stream;
+use futures::sync::mpsc::channel;
+use rand::Rng;
+use tokio::timer::Delay;
+
+use crate::protocol::Package;
+use crate::proxy::ProxyHandler;
+
+/// Messages passing through this handler will be delayed by a random number of milliseconds
+/// in the range `[0, max_delay_ms)`
+pub struct ThrottlingHandler {
+    max_delay_ms: u64
+}
+
+impl ThrottlingHandler {
+    pub fn new(max_delay_ms: u64) -> Self {
+        Self {
+            max_delay_ms
+        }
+    }
+}
+
+/// Messages will be dropped with probability `dropout_rate `
+impl ProxyHandler for ThrottlingHandler {
+    fn pipe_stream(&self, stream: Box<Stream<Item=Package, Error=()> + Send + Sync>) ->
+    Box<Stream<Item=Package, Error=()> + Send + Sync>
+    {
+        let (message_tx, message_rx) = channel(1024);
+
+        let max_delay_ms = self.max_delay_ms;
+
+        let main_task = stream.for_each(move |package| {
+            let mut rng = rand::thread_rng();
+            let delay = (rng.gen::<f64>() * max_delay_ms as f64) as u64;
```

Comment:
I don't think `rng.gen::<f64>()` is in [0,1] range.

Reply:
It is on range [0,1) [reference](https://docs.rs/rand/0.6.5/rand/distributions/struct.Standard.html#built-in-implementations)!